Add all color palettes to select from editor panel

AKSHAT2802 opened the pull request:
Closes: https://github.com/WordPress/gutenberg/issues/64857
Closes: https://github.com/WordPress/gutenberg/issues/64734
## What?
This PR Adds all the available colors in the list of editor color panel ( default, theme, custom )  so that the user can any predefined color in the editor.

## Why?
When selecting a color we should have a list of all the predefined options for a user instead of only custom-defined options.

## How?
Modified with-color-context.js file to list all the color instead of only custom colors.

## Screenshots or screencast
### Before
<img width="705" alt="Screenshot 2024-09-09 at 5 07 46 PM" src="https://github.com/user-attachments/assets/b2d7d9a1-de81-4671-9cda-8c44ca26bc6f">

### After

https://github.com/user-attachments/assets/895492b8-0601-461d-affb-cd6361c5ae5e

AKSHAT2802 committed:
Add all the colors in with-color-context so that they all are listed in block editor

ndiego commented:
Thanks for working on this @AKSHAT2802. However, I don't think we should display the default color palette as well, only the theme color palette in addition to any custom options set by the user. Using the Twenty Twenty-Four theme as an example, this is what it should look like. 

|  There are no custom colors | The user has added custom colors |
|-|-|
|<img width="668" alt="image" src="https://github.com/user-attachments/assets/a65660eb-43d0-4e56-84e3-6338494837d7">|<img width="698" alt="image" src="https://github.com/user-attachments/assets/1959b083-d58a-4e4a-a245-db80643461ac">|

richtabor reviewed (changes requested):
Two notes: 

1. I should only see the palettes which I have enabled. If I don't have the default palette enabled via theme.json, then I should not see that palette here.

2. I'd expect the colors in the same order as the color palette elsewhere, with theme colors first, then custom colors. In the screenshot below I'd expect to see my theme colors, then the custom red color: 
![CleanShot 2024-11-07 at 09 19 10](https://github.com/user-attachments/assets/63a81b21-ed74-427c-bbbc-35c38d193451)

AKSHAT2802 committed:
Merge branch 'trunk' into feat/display-all-color-palette-in-editor

AKSHAT2802 committed:
Remove default colour pallete

AKSHAT2802 committed:
Remove extra space

AKSHAT2802 commented:
Hii @richtabor & @ndiego 
Thanks for the review 
Updated the PR

Sample Images 

1] Theme palette with custom colors
![image](https://github.com/user-attachments/assets/13db0ef8-b162-4878-b3b3-7c6be2f5093a)

2] Changed theme palette with no custom colors
![image](https://github.com/user-attachments/assets/191bfc4c-2e02-4371-ac4e-f76e30493a38)

3] Changed theme palette and added custom colors
![image](https://github.com/user-attachments/assets/f8268e7c-8c4b-4e22-a7da-5172876b515f)

t-hamano commented on packages/block-editor/src/components/color-palette/with-color-context.js:
```diff
@@ -10,14 +10,17 @@ import { useSettings } from '../use-settings';
-			disableCustomColors = ! enableCustomColors,
-		} = props;
+
+		const colors = [ ...( themeColors || [] ), ...( customColors || [] ) ];
```
```suggestion
		const colors = [ ...themeColors, ...defaultColors, ...customColors ];
```

This simplification should be fine.

t-hamano reviewed (commented):
I think the default palette should not be disabled all the time, but only when `settings.color.defaultPalette` is `false`, otherwise it won't match other color palettes. Moreover, the order is "Theme" > "Default" > "Custom":

## Current

![image](https://github.com/user-attachments/assets/a48097cb-897d-4e6f-b7dc-f5e447949761)

## Expected

![image](https://github.com/user-attachments/assets/20cc76d9-8147-44af-b293-11a63a98988f)

Additionally, don't forget to test the inline text highlighting palette in this PR:

![image](https://github.com/user-attachments/assets/9c97b3fe-845a-4e75-a4ac-e1e46913e9c1)

richtabor commented:
> I think the default palette should not be disabled all the time, but only when settings.color.defaultPalette is false, otherwise it won't match other color palettes. Moreover, the order is "Theme" > "Default" > "Custom":

+1

AKSHAT2802 committed:
Update packages/block-editor/src/components/color-palette/with-color-context.js

re-add default colours

Co-authored-by: Aki Hamano <54422211+t-hamano@users.noreply.github.com>

AKSHAT2802 committed:
Add default colours if enabled

AKSHAT2802 commented:
Thanks, @t-hamano for the update. 
I have changed the code according to the requirements.

t-hamano commented on packages/block-editor/src/components/color-palette/with-color-context.js:
```diff
@@ -10,14 +10,27 @@ import { useSettings } from '../use-settings';
-			disableCustomColors = ! enableCustomColors,
-		} = props;
+
+		const colors = enableDefaultColors
+			? themeColors.concat( defaultColors, customColors )
+			: themeColors.concat( customColors );
+
+		const { disableCustomColors = ! enableCustomColors } = props;
```
```suggestion
		const _colors = enableDefaultColors
			? [
					...( themeColors || [] ),
					...( defaultColors || [] ),
					...( customColors || [] ),
			  ]
			: [ ...( themeColors || [] ), ...( customColors || [] ) ];

		const { colors = _colors, disableCustomColors = ! enableCustomColors } =
			props;
```

This change contains two fixes:

- Fixes a critical error that occurs when theme colors or custom colors are missing.
- Respects the `colors` prop of the `ColorPalette` component.

AKSHAT2802 committed:
Update packages/block-editor/src/components/color-palette/with-color-context.js

Co-authored-by: Aki Hamano <54422211+t-hamano@users.noreply.github.com>

Commit:
update test

richtabor reviewed (approved):
LGTM 
![CleanShot 2024-11-21 at 17 48 04](https://github.com/user-attachments/assets/8305c42d-68b0-4482-979e-b5394db78dde)

t-hamano commented:
It has passed all CI checks so let's merge this PR.

@AKSHAT2802 Thank you for your efforts!
